java-frontend/src/main/java/org/sonar/java/model/LiteralUtils.java

```diff
@@ -148,4 +150,28 @@ public static boolean isOne(ExpressionTree tree) {
   public static boolean isNegOne(ExpressionTree tree) {
     return tree.is(Kind.UNARY_MINUS) && isOne(((UnaryExpressionTree) tree).expression());
   }
+  
+  public static String getAsStringValue(LiteralTree tree) {
+    if (!tree.is(Kind.TEXT_BLOCK)) {
+      return trimQuotes(tree.value());
+    }
+    String[] lines = trimQuotes(tree.value()).split("\r?\n");
+    int indent = Arrays.stream(lines).skip(1).mapToInt(LiteralUtils::getIndentation)
+      .min().orElse(0);
+
+    return Arrays.stream(lines)
+      .skip(1)
+      .map(s -> s.substring(indent))
+      .map(s -> s.replaceAll("[\t ]*$", ""))
+      .collect(Collectors.joining("\n"));
+  }
+  
+  private static int getIndentation(String str) {
+    for (int i = 0; i < str.length(); ++i) {
+      if (str.charAt(i) != ' ' && str.charAt(i) != '\t' ) {
```

Comment:
Another thing I missed the first time (and in my own rule implementation as well): Form feeds (`\f`) are also considered as space characters - both for the purposes of counting indentation and stripping trailing spaces. That isn't a really a big deal because how would a form feed even end up in a Java file? But if you're going to reopen this issue anyway, you might as well add a `\f` here and on the `replaceAll` call.

That's it though: spaces, tabs and form feeds. Unicode spaces count as regular characters - so good luck figuring out why your indentation isn't stripped if one of your indentation spaces happens to be a Unicode space.